packages/neuron-wallet/src/services/addresses.ts

```diff
@@ -149,128 +160,196 @@ export default class AddressService {
+      .filter(addr => addr.addressType === AddressType.Change)
+
+    return [unusedReceivingAddresses, unusedChangeAddresses]
   }
 
-  public static nextUnusedChangeAddress = (walletId: string): AddressInterface | undefined => {
-    const addressEntity = AddressDao.nextUnusedChangeAddress(walletId,  AddressService.getAddressVersion())
-    if (!addressEntity) {
-      return undefined
+  public static async getAllUnusedAddressesByWalletId (walletId: string) {
+    const txCountsByLockHashes = await TransactionsService.getTxCountsByWalletId(walletId)
+
+    const addresses = await this.allAddressesByWalletId(walletId)
+    for (const address of addresses) {
+      const script = Script.fromObject({
+        codeHash: SystemScriptInfo.SECP_CODE_HASH,
+        hashType: SystemScriptInfo.SECP_HASH_TYPE,
+        args: address.blake160
+      });
+      const lockHash = script.computeHash()
+      const txCount = txCountsByLockHashes.get(lockHash)
+
+      if (!txCount) {
+        continue
+      }
+      address.txCount = txCount
     }
-    return addressEntity
+
+    const allUnusedAddresses = addresses.filter(address => !address.txCount)
+
+    return allUnusedAddresses
   }
 
-  public static allAddresses = (): AddressInterface[] => {
-    return AddressDao.allAddresses( AddressService.getAddressVersion())
+  public static async nextUnusedAddress (walletId: string): Promise<AddressInterface | undefined> {
+    const [unusedReceivingAddresses, unusedChangeAddresses] = await this.getUnusedAddresses(walletId)
+    if (unusedReceivingAddresses.length) {
+      return unusedReceivingAddresses[0]
+    }
+    if (unusedChangeAddresses.length) {
+      return unusedChangeAddresses[0]
+    }
+
+    return undefined
   }
 
-  public static allAddressesByWalletId = (
-    walletId: string,
-    addressVersion: AddressVersion = AddressService.getAddressVersion()
-  ): AddressInterface[] => {
-    return AddressDao.allAddressesByWalletId(walletId, addressVersion)
+  public static async nextUnusedChangeAddress (walletId: string): Promise<AddressInterface | undefined> {
+    const [ , unusedChangeAddresses ] = await this.getUnusedAddresses(walletId)
+    if (unusedChangeAddresses.length) {
+      return unusedChangeAddresses[0]
+    }
+    return undefined
   }
 
-  public static allBlake160sByWalletId(walletId: string): string[] {
-    return AddressService.allAddressesByWalletId(walletId).map(addr => addr.blake160)
+  public static async allUnusedReceivingAddresses(walletId: string): Promise<AddressInterface[]> {
+    const [ unusedReceivingAddresses ] = await this.getUnusedAddresses(walletId)
+    return unusedReceivingAddresses
   }
 
-  public static allLockHashes(): string[] {
-    const addresses = AddressService.allAddresses().map(address => address.address)
-    return AddressParser.batchToLockHash(addresses)
+
+  public static allAddresses = async (): Promise<AddressInterface[]> => {
+    const publicKeyInfos = await AddressDao.allAddresses()
```

Comment:
The name is `allAddresses` but only those of `AddressDao` are returned?

Reply:
The name of `AddressDao` is quite misleading. It encapsulating the data operations for addresses no matter it is related to DAO or not. I will merge this `AddressDao` class into `AddressService`.